src/handlers/proposalsHandlers.ts

```diff
@@ -56,7 +62,175 @@ const getProposals = (
     });
 };
 
-const getProposal = (
+/**
+ * Merges the given field values into a new copy of the given proposal.
+ * It is OK if there are no versions or zero field values.
+ * @param proposal The proposal with versions but no values.
+ * @param values The values to merge into the given proposal.
+ * @returns A new proposal with the values merged into their respective versions.
+ */
+export const mergeProposalFieldValues = (
```

Comment:
Instead of merging field values onto proposals we could have a simpler function that merges field values onto proposal VERSIONS.  This would have the added benefit of being more universally relevant (and would / could go into a `/utils` and be used on other handlers that involve deeply populated proposal versions).

I think this would result in a dramatically less complex operation; you would then be able to do something like:

```
proposalVersions.map((proposalVersion) => decorateProposalVersionWithFieldValues(
  proposalVersion,
  proposalFieldValues,
));
```

^ and you could have the decorator just skip any field values whose proposalVersionId doesn't align with the proposal version.

Your deep lookup function could then be something as simple as this:

```typescript
Promise.all(
  proposalQuery,
  proposalVersionQuery,
  proposalFieldValuesQuery,
).then(([proposals, proposalVersions, proposalFieldValues]) => {
  const decoratedProposalVersions = proposalVersions.map(
    (proposalVersion) => decorateProposalVersionWithFieldValues(
      proposalVersion,
      proposalFieldValues,
    )
  );
  const decoratedProposals = proposals.map((proposal) => decorateProposalWithVersions(
    proposal,
    decoratedProposalVersions,
  ))
  res.send(decoratedProposals) // or whatever
})
```

Reply:
So then you are iterating over the proposal field values (having count N) for each proposal version (having count M), so N*M instead of N. Such N*M iteration is what the comment below was trying to communicate that we avoid. I think your suggestion or something like it isl worth doing because it truly is simpler, more targeted, etc., but there will need to be a separate step of getting the proposal versions split up or accepting M queries per request (yuck), or I suppose letting the iterations happen because the expensive part is over (e.g. expensive to get from a disk, we're merely iterating over heap memory objects now more cheaply).

Reply:
Yeah I think we still stick with JUST the three queries (getting ALL data at once).

We could then have something like `filterFieldValuesByProposalVersionId` that just filters out the ones we want -- which I know is slightly more wasteful in runtime since it would filter the array N * M times, but I also don't think we need to optimize to that level at this point in the pilot (and to your point the thing to really avoid is hitting disk / running a query for each)

If we wanted to optimize (but this would get a little more code complex) then we could also pre-process the query result with a generic utility like "generateGroupsByAttribute" which would return an object of `Record<attributesType, T[]>`

```
generateGroupsByAttribute<T>(items: T[], attribute: string) => items.reduce(
  (groups, item) => {
    if (!item.hasOwnProperty(attribute)) {
      return groups;
    }
    return {
     ...groups,
     [item[attribute]]: [
       ...(groups[item[attribute]] ?? []),
       item
  }),
  {},
);
```

^ this (untested, written in github) function would split an array into groups with keys of whatever attribute.  So you could just do something like this:

```
  const groupedProposalFieldValues = generateGroupsByAttribute(proposalFieldValues, 'proposalVersionId');
  const decoratedProposalVersions = proposalVersions.map(
    (proposalVersion) => decorateProposalVersionWithFieldValues(
      proposalVersion,
      groupedProposalFieldValues[proposalVersion.id] ?? [],
    );
  );
```

^^ Again that's a bit more complex (though O(N)) than something like: 

```
  const decoratedProposalVersions = proposalVersions.map(
    (proposalVersion) => decorateProposalVersionWithFieldValues(
      proposalVersion,
      groupedProposalFieldValues.filter((proposalFieldValue) => proposalFieldValue.proposalVersionId === proposalVersion.id)
    );
  );
```

Reply:
I tend not to genericize in advance but rather after a thing is used more than once. The whole discussion or review surrounds code that we expect to be copied/pasted so I think your anticipation of re-use is well-founded.

Reply:
100% agree with the general "don't-over-generalize" philosophy!  +1 to the idea that we can anticipate it to some degree here so there's a balance worth striking!

Reply:
The reason for the more fine-grained `await` calls was also perhaps premature optimization. The idea is like this: it is safe to guess that selecting one proposal row will be quicker than selecting M proposal versions rows which will be quicker than selecting N proposal field values rows (and N application form fields rows). After the first two queries complete, there is no reason we can't do the `join` or mash those together while waiting for the remaining two to complete. `Promise.all` will block until the slowest query completes. That said, your suggestion is much easier to reason about when reading the code so I will try it. It will be slightly more complex with the Ajv validations.  I am not sure about whether the types will be preserved either, I will try it and see.

Reply:
The merge that happens first uses the two biggest query results after all. `Promise.all` it is!